keccak256/src/permutation/tables.rs

```diff
@@ -486,202 +489,257 @@ impl<F: Field> Base13toBase9TableConfig<F> {
+        let lookup_config = ThreeColumnsLookup::configure(meta, adv_cols, table_cols, "from base9");
+        let map = HashMap::new();
+        Self { lookup_config, map }
     }
+    pub fn assign_region(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        input: &AssignedCell<F, F>,
+    ) -> Result<
+        (
+            Vec<AssignedCell<F, F>>,
+            Vec<AssignedCell<F, F>>,
+            Vec<AssignedCell<F, F>>,
+        ),
+        Error,
+    > {
+        let input_coefs = compute_input_coefs::<F, NUM_OF_B9_SLICES>(
+            input.value(),
+            B9,
+            NUM_OF_B9_CHUNKS_PER_SLICE,
+        );
+        layouter.assign_region(
+            || "base 9",
+            |mut region| {
+                let mut input_cells = vec![];
+                let mut output_b13_cells = vec![];
+                let mut output_b2_cells = vec![];
+                for (offset, input_coef) in input_coefs.iter().enumerate() {
+                    self.lookup_config.q_enable.enable(&mut region, offset)?;
+                    let input = region.assign_advice(
+                        || "base 9",
+                        self.lookup_config.col0.0,
+                        offset,
+                        || input_coef.ok_or(Error::Synthesis),
+                    )?;
+                    input_cells.push(input.clone());
+                    let output = input_coef.and_then(|v| self.map.get(&v.to_repr()));
 
-    pub(crate) fn get_base_info(&self, output_b9: bool) -> BaseInfo<F> {
-        BaseInfo {
-            input_base: B2,
-            output_base: if output_b9 { B9 } else { B13 },
-            num_chunks: NUM_OF_BINARY_CHUNKS,
-            max_chunks: MAX_CHUNKS,
-            input_tc: self.base2,
-            output_tc: if output_b9 { self.base9 } else { self.base13 },
-            _marker: PhantomData,
-        }
+                    let output_b13 = region.assign_advice(
+                        || "base 13",
+                        self.lookup_config.col1.0,
+                        offset,
+                        || output.map(|v| v.0).ok_or(Error::Synthesis),
+                    )?;
+                    output_b13_cells.push(output_b13);
+                    let output_b2 = region.assign_advice(
+                        || "base 2",
+                        self.lookup_config.col2.0,
+                        offset,
+                        || output.map(|v| v.1).ok_or(Error::Synthesis),
+                    )?;
+                    output_b2_cells.push(output_b2);
+                }
+                Ok((input_cells, output_b13_cells, output_b2_cells))
+            },
+        )
     }
 }
 
 #[derive(Debug, Clone)]
-pub struct FromBase9TableConfig<F> {
-    base9: TableColumn,
-    base13: TableColumn,
-    base2: TableColumn,
-    _marker: PhantomData<F>,
+pub struct FromBinaryTableConfig<F> {
+    lookup_config: ThreeColumnsLookup<F>,
+    // mapping from base9 input to base13 and base2 output
```

Comment:
If the input comes from base9, why is the struct called `FromBinaryTableConfig`? That would make sense if the comment stated b2 -> (b9 & b13)

Reply:
I guess seeing the implementation that the comment is wrong and it should be `base2` input to `base13 and base9` output

Reply:
Nice catch